### ISSUE TITLE
Darwin shows `install: skipping file '/dev/fd/63', as it was replaced while being copied` when building derivations

### ISSUE THREAD

**Issue by smancill** [NixOS/nixpkgs#335016]
### Describe the bug

I started seeing this message when building derivations on Darwin:

```
install: skipping file '/dev/fd/63', as it was replaced while being copied
```

Multiple times, in multiple phases.

It also happens in ofborg. See for example: https://github.com/NixOS/nixpkgs/pull/334542/checks?check_run_id=28783170448

### Steps To Reproduce

- Build a derivation on Darwin. It happens with any derivation as I have been doing multiple `nixpkg-review`s

### Expected behavior

There should not be any error message related to `install`. It may be innocuous so far but I'd rather not see errors.

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-darwin"`
 - host os: `Darwin 20.6.0, macOS 10.16`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.5`
 - channels(smancill): `"darwin, nixpkgs-23.11-darwin, unstable"`
 - nixpkgs: `/Users/smancill/.nix-defexpr/channels/nixpkgs`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by s0me1newithhand7s** [comment]
isn't **Metadata** part cruical?

**Comment by smancill** [comment]
> isn't **Metadata** part cruical?

Yes, I forgot even when I did see it.

**Comment by n8henrie** [comment]
Also started seeing this today.

```console
$ nix-info -m
 - system: `"aarch64-darwin"`
 - host os: `Darwin 23.6.0, macOS 14.6.1`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.5`
 - channels(n8henrie): `""`
 - channels(root): `""`
 - nixpkgs: `/nix/store/pxzmw3l4kzswvfgyjlxxdx35vchxx3ns-source`
```

cc @NixOS/darwin-maintainers

**Comment by afh** [comment]
Thanks for raising this, @smancill, what is the "smallest" or most simple package that you've come across, that exhibits the described misbehaviour?

**Comment by reckenrode** [comment]
Probably the same issue behind https://github.com/NixOS/nixpkgs/pull/300797.

**Comment by smancill** [comment]
> Thanks for raising this, @smancill, what is the "smallest" or most simple package that you've come across, that exhibits the described misbehaviour?

`cmark` is a fairly simple package.  Also GNU `hello` I'd say, but the configure call is a bit slow.

**Comment by smancill** [comment]
As I understand the file descriptor `/dev/fd/63` is for process substitution: `$(cmd)`.

**Comment by viraptor** [comment]
So the merge that changed it was commit 9ccfa877218074582ff21fe950944fa0a7240248
I'm just following with a bisect of that branch.

**Comment by viraptor** [comment]
It's commit c47a1e701df7f00352b8cf401fa79c0d2f5fcc59
Pinging @LeSuisse in case he has an idea.

**Comment by n8henrie** [comment]
I get the same error message from above when using (?GNU) install from nix and a different one from (?BSD) install:

```console
$ type -a install
install is /etc/profiles/per-user/n8henrie/bin/install
install is /run/current-system/sw/bin/install
install is /usr/bin/install
$
$ install -m 0600 <(export) /tmp/foo.install
install: skipping file '/dev/fd/63', as it was replaced while being copied
$
$ /usr/bin/install -m 0600 <(export) /tmp/foo.install
install: /dev/fd/63: Inappropriate file type or format
```

**Comment by LeSuisse** [comment]
Great...

Could some of you test with the following change?

```diff
diff --git a/pkgs/stdenv/generic/setup.sh b/pkgs/stdenv/generic/setup.sh
index fc748aa12a66..4b32ffb6cd8f 100644
--- a/pkgs/stdenv/generic/setup.sh
+++ b/pkgs/stdenv/generic/setup.sh
@@ -1093,7 +1093,8 @@ substituteAllInPlace() {
 # the environment used for building.
 dumpVars() {
     if [ "${noDumpEnvVars:-0}" != 1 ]; then
-        install -m 0600 <(export 2>/dev/null) "$NIX_BUILD_TOP/env-vars" || true
+        install -m 0600 /dev/null "$NIX_BUILD_TOP/env-vars" || true
+        export 2>/dev/null >| "$NIX_BUILD_TOP/env-vars" || true
     fi
 }
```

**Comment by viraptor** [comment]
I think we can replace this with:
```
save_umask=$(umask)
umask 077
(the old command)
umask $save_umask
```
Testing it now.

**Comment by n8henrie** [comment]
What is the reason behind the `|| true` here? Shouldn't this just fail -- leading to us potentially having caught this much earlier / perhaps prior to merge?

**Comment by LeSuisse** [comment]
> What is the reason behind the `|| true` here? Shouldn't this just fail -- leading to us potentially having caught this much earlier / perhaps prior to merge?

I kept it as close as it was initially, I'm guessing it was made this way because it is a debugging thing and it should not be a big deal if it breaks.

**Comment by viraptor** [comment]
@LeSuisse your idea works just fine for me 👍

**Comment by LeSuisse** [comment]
Created #335593 with the change, thanks for the confirmation :+1:

**Comment by reckenrode** [comment]
> It looks like `install` on linux is fine with this idea, but macos is not:

https://debbugs.gnu.org/cgi/bugreport.cgi?bug=70411

Someone sent a feedback to Apple, but who knows if/when the underlying issue in XNU will be fixed.

**Comment by NfNitLoop** [comment]
New user here, seeing this on the first commands I run with nix-shell. 

```
> nix-shell -p cowsay --run "cowsay moo"
install: skipping file '/dev/fd/63', as it was replaced while being copied
 _____
< moo >
 -----
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||
```

**Comment by avanov** [comment]
Seeing this too after bumping my pins to the recent https://github.com/NixOS/nixpkgs/commit/e9b5094b8f6e06a46f9f53bb97a9573b7cedf2a2 from the unstable channel.

**Comment by emilazy** [comment]
We have a fix in the works but unfortunately it will have to go through the staging build cycle, so it’ll be a few weeks before the message goes away. It should be harmless though.

**Comment by smancill** [comment]
Fixed by #335593
